File: Рудович/1483.cpp
#include <vector>
#include <cmath>

using namespace std;

class TreeAncestor {
private:
vector<vector<int>> dp;

public:
    TreeAncestor(int n, vector<int>& parent): dp(n, vector<int>(log2(n) + 1, -1))
    {
        for (int i = 0; i < n; ++i)
        {
            fun(parent, i);
        }
    }

    void fun(vector<int>& parent, int k)
    {
        if (parent[k] == -1)
        {
            return;
        }
        if (dp[k][0] != -1)
        {
            return;
        }
        dp[k][0] = parent[k];
        for (int i = 1; i < dp[k].size(); ++i)
        {
            if (dp[k][i - 1] == -1)
            {
                dp[k][i] = -1;
            }
            else
            {
                fun(parent, dp[k][i - 1]);
                dp[k][i] = dp[dp[k][i - 1]][i - 1];
            }
        }
    }
    
    int getKthAncestor(int node, int k) 
    {
        if (node == -1)
        {
            return -1;
        }
        if (k == 0)
        {
            return node;
        }
        int p = 0;
        int s = 1;
        while(s * 2 < k)
        {
            ++p;
            s *= 2;
        }
        return getKthAncestor(dp[node][p], k - s);
    }
};

File: Матюхин/1483.ts
class TreeAncestor {
    private sparseTable: number[][];

    private dfsSparceTable(adjacentVertices: Set<number>[]): void {
        const reachArray = new Array(adjacentVertices.length).fill(false);
        const dfsStack = [0];

        while (dfsStack.length !== 0) {
            const currentVertex = dfsStack.pop()!;

            if (!reachArray[currentVertex]) {
                for (let j = 1; j < this.sparseTable.length; ++j) {
                    const aboveRow = this.sparseTable[j - 1];
                    const aboveVertex = aboveRow[currentVertex];
                    const currentRow = this.sparseTable[j];

                    if (aboveVertex !== -1) {
                        // Assign 2^j-th ancestor to the current position.
                        currentRow[currentVertex] = aboveRow[aboveVertex];
                    }
                }
            }

            for (const vertex of adjacentVertices[currentVertex]) {
                if (!reachArray[vertex]) {
                    reachArray[currentVertex] = true;
                    dfsStack.push(vertex);
                }
            }
        }
    }

    constructor(_: number, parentNumbers: number[]) {
        const adjacentVertices = Array.from(
            { length: parentNumbers.length },
            () => new Set<number>(),
        );

        this.sparseTable = [[...parentNumbers], ...Array.from({
            length: Math.floor(Math.log2(parentNumbers.length)) + 1,
        }, () => new Array(parentNumbers.length).fill(-1))];

        parentNumbers.forEach((parentNumber, vertex) => {
            if (parentNumber !== -1) {
                adjacentVertices[vertex].add(parentNumber);
                adjacentVertices[parentNumber].add(vertex);
            }
        });

        this.dfsSparceTable(adjacentVertices);
    }

    getKthAncestor(node: number, k: number): number {
        while (k !== 0) {
            if (node === -1) {
                return -1;
            }

            const kPower2 = Math.floor(Math.log2(k));
            node = this.sparseTable[kPower2][node];
            k -= 2 ** kPower2;
        }

        return node;
    }
}
